src/validation/input-params.ts

```diff
@@ -330,85 +234,73 @@ class ProcessedParam<const T extends InputParameter = InputParameter> {
           throw this.validationError('input value must be a non-empty array')
-        } else {
-          return []
         }
+        return []
       }
 
-      // Validate each value from the array individually
       return input.map((item) => this.validateInputType(item))
     }
 
-    // We already know this won't be an array, but the generics are too complex for typescript
-    // to infer by itself so we cast manually
     return this.validateInputType(input as ShouldBeUndefinable<T>)
   }
 
-  /**
-   * Validates a single value from the incoming params.
-   *
-   * @param input - a single value from the request params object
-   * @returns - the validated input data
-   */
   private validateInputType(input: NonArrayInputType<T>) {
-    // If we're here we've already checked this is not required
     if (input == null) {
       return
     }
 
-    // If the type is a nested input params object, use that to validate
     if (this.type instanceof InputParameters) {
       return this.type.validateInput(input)
     }
 
-    // If the param has specified options, check that the input is one of them.
-    // In this case we don't need to check the type, since the options will do that for us
     if (this.options && !this.options.has(input as TypeFromParameter<T>)) {
       throw this.validationError(`input is not one of valid options (${this.definition.options})`)
-    } else if (typeof input !== this.definition.type) {
-      throw this.validationError(`input type is not the expected one (${this.type})`)
+    }
+    if (typeof input !== this.definition.type) {
+      throw this.validationError(`input type is not the expected one (${this.definition.type})`)
     }
 
-    // If no validations failed and no defaults / modifications were applied, use the original input
     return input
   }
 }
 
+/* -------------------------------------------------------------------------- */
+/*                 MAIN CLASS – WITH COMPILED AJV VALIDATOR                   */
+/* -------------------------------------------------------------------------- */
+
 export class InputParameters<const T extends ProperInputParametersDefinition> {
-  // Helper type to avoid doing TypeFromDefinition<typeof inputParameters.definition>
-  // Should not be used in practice
+  /* Helper type; not emitted at runtime */
   readonly validated!: TypeFromDefinition<T>
 
   params: ProcessedParam[]
 
+  private readonly schema: AnySchema
+  private readonly validateFn: ValidateFunction
+
   constructor(
     public definition: T,
     public examples?: TypeFromDefinition<T>[],
   ) {
     this.params = Object.entries(this.definition).map(
       ([name, param]) => new ProcessedParam(name, param),
     )
-
-    // Check that all options match param type
-    // Check that defaults matches param validation
     this.validateDefinition()
+
+    /* New: compile schema once */
+    this.schema = definitionToJsonSchema(this.definition)
+    this.validateFn = getValidator(this.schema)
```

Comment:
We tried ajv fast too, but at the time the parameters were very complex and proved impossible to do the migrations properly so it was quickly rejected. Still, it's a big time sink on the read request execution, so any speedup helps a lot and worth a try. I would first consider the approach where the validation is skipped entirely for mercury jobs though, as we avoid this altogether. Paired with the suggestion to precalculate the cacheKeys, requests could be resolved in 2 lines only and be blazing fast